# HG changeset patch
# User Olaf Till <i7tiol@t-online.de>
# Date 1332464452 14400
#      Thu Mar 22 21:00:52 2012 -0400
# Node ID 7ce925166af650c6470653f17ae5f7ee13c6ad19
# Parent  5bd9e47e927744401310f4582d25c6664dbc0638
Fix inconsistent orientation of output matrix for interp1.

* interp1.m: Fix inconsistent orientation of output matrix. Properly include
the extrapolaed values.

diff --git a/scripts/general/interp1.m b/scripts/general/interp1.m
--- a/scripts/general/interp1.m
+++ b/scripts/general/interp1.m
@@ -243,31 +243,39 @@ function yi = interp1 (x, y, varargin)
   case {"pchip", "*pchip", "cubic", "*cubic"}
     if (nx == 2 || starmethod)
       x = linspace (x(1), x(nx), ny);
     endif
 
     if (ispp)
       y = shiftdim (reshape (y, szy), 1);
       yi = pchip (x, y);
+      yi.orient = "first";
     else
       y = shiftdim (y, 1);
       yi = pchip (x, y, reshape (xi, szx));
+      if (! isvector (y))
+        yi = shiftdim (yi, 1);
+      endif
     endif
   case {"spline", "*spline"}
     if (nx == 2 || starmethod)
       x = linspace(x(1), x(nx), ny);
     endif
 
     if (ispp)
       y = shiftdim (reshape (y, szy), 1);
       yi = spline (x, y);
+      yi.orient = "first";
     else
       y = shiftdim (y, 1);
       yi = spline (x, y, reshape (xi, szx));
+      if (! isvector (y))
+        yi = shiftdim (yi, 1);
+      endif
     endif
   otherwise
     error ("interp1: invalid method '%s'", method);
   endswitch
 
   if (! ispp)
     if (! ischar (extrap))
       ## determine which values are out of range and set them to extrap,
@@ -275,30 +283,21 @@ function yi = interp1 (x, y, varargin)
       minx = min (x(1), x(nx));
       maxx = max (x(1), x(nx));
 
       outliers = xi < minx | ! (xi <= maxx); # this catches even NaNs
       if (size_equal (outliers, yi))
         yi(outliers) = extrap;
         yi = reshape (yi, szx);
       elseif (!isvector (yi))
-        if (strcmp (method, "pchip") || strcmp (method, "*pchip")
-          ||strcmp (method, "cubic") || strcmp (method, "*cubic")
-          ||strcmp (method, "spline") || strcmp (method, "*spline"))
-          yi(:, outliers) = extrap;
-          yi = shiftdim(yi, 1);
-        else
-          yi(outliers, :) = extrap;
-        endif
+        yi(outliers, :) = extrap;
       else
         yi(outliers.') = extrap;
       endif
     endif
-  else
-    yi.orient = "first";
   endif
 
 endfunction
 
 
 %!demo
 %! clf;
 %! xf = 0:0.05:10;  yf = sin (2*pi*xf/5);
